ml/set_upgrade_mode.yml#"Setting upgrade_mode to enabled" failure on CI

Failed on 7.0 intake: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.0+intake/500/console (new intake commit is an unrelated docs change)

Reproduce line (does not reproduce locally):
```
./gradlew :x-pack:plugin:integTestRunner \
  -Dtests.seed=971A963D9AC18AC0 \
  -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT \
  -Dtests.method="test {p0=ml/set_upgrade_mode/Setting upgrade_mode to enabled}" \
  -Dtests.security.manager=true \
  -Dtests.locale=lt \
  -Dtests.timezone=Europe/Lisbon \
  -Dcompiler.java=12 \
  -Druntime.java=8
```

Stack trace:
```
java.lang.RuntimeException: Had to resort to force-closing jobs, something went wrong?
	at __randomizedtesting.SeedInfo.seed([971A963D9AC18AC0:1F4EA9E7343DE738]:0)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:85)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.clearMlMetadata(MlRestTestStateCleaner.java:31)
	at org.elasticsearch.xpack.test.rest.XPackRestIT.clearMlState(XPackRestIT.java:270)
	at org.elasticsearch.xpack.test.rest.XPackRestIT.cleanup(XPackRestIT.java:254)
	at sun.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.elasticsearch.client.ResponseException: method [POST], host [http://[::1]:38309], URI [/_ml/anomaly_detectors/_all/_close], status line [HTTP/1.1 500 Internal Server Error]
{"error":{"root_cause":[{"type":"exception","reason":"Unable to read field length from input stream\nFailed to parse length encoded header from stream\nFailed to handle all input data\nMl anomaly detector job failed\n"}],"type":"exception","reason":"Exception closing autodetect process","caused_by":{"type":"exception","reason":"Unable to read field length from input stream\nFailed to parse length encoded header from stream\nFailed to handle all input data\nMl anomaly detector job failed\n"}},"status":500}
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:260)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:238)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:212)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:77)
	... 37 more
```

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectCommunicator.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectCommunicator.java
@@ -90,9 +90,8 @@ public class AutodetectCommunicator implements Closeable {
                 && job.getAnalysisConfig().getCategorizationFieldName() != null;
     }
 
-    public void init(ModelSnapshot modelSnapshot) throws IOException {
+    public void restoreState(ModelSnapshot modelSnapshot) {
         autodetectProcess.restoreState(stateStreamer, modelSnapshot);
-        createProcessWriter(Optional.empty()).writeHeader();
     }
 
     private DataToProcessWriter createProcessWriter(Optional<DataDescription> dataDescription) {
@@ -101,6 +100,17 @@ public class AutodetectCommunicator implements Closeable {
                 dataCountsReporter, xContentRegistry);
     }
 
+    /**
+     * This must be called once before {@link #writeToJob(InputStream, AnalysisRegistry, XContentType, DataLoadParams, BiConsumer)}
+     * can be used
+     */
+    public void writeHeader() throws IOException {
+        createProcessWriter(Optional.empty()).writeHeader();
+    }
+
+    /**
+     * Call {@link #writeHeader()} exactly once before using this method
+     */
     public void writeToJob(InputStream inputStream, AnalysisRegistry analysisRegistry, XContentType xContentType,
                            DataLoadParams params, BiConsumer<DataCounts, Exception> handler) {
         submitOperation(() -> {

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
@@ -434,7 +434,6 @@ public class AutodetectProcessManager implements ClusterStateListener {
                             return;
                         }
 
-                        processByAllocation.putIfAbsent(jobTask.getAllocationId(), new ProcessContext(jobTask));
                         jobResultsProvider.getAutodetectParams(job, params -> {
                             // We need to fork, otherwise we restore model state from a network thread (several GET api calls):
                             threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(new AbstractRunnable() {
@@ -445,7 +444,8 @@ public class AutodetectProcessManager implements ClusterStateListener {
 
                                 @Override
                                 protected void doRun() {
-                                    ProcessContext processContext = processByAllocation.get(jobTask.getAllocationId());
+                                    ProcessContext processContext = new ProcessContext(jobTask);
+
                                     if (processContext == null) {
                                         logger.debug("Aborted opening job [{}] as it has been closed", jobId);
                                         return;
@@ -460,6 +460,9 @@ public class AutodetectProcessManager implements ClusterStateListener {
                                         createProcessAndSetRunning(processContext, job, params, closeHandler);
                                         processContext.getAutodetectCommunicator().restoreState(params.modelSnapshot());
                                         setJobState(jobTask, JobState.OPENED);
+
+                                        assert processByAllocation.get(jobTask.getAllocationId()) == null;
+                                        processByAllocation.put(jobTask.getAllocationId(), processContext);
                                     } catch (Exception e1) {
                                         // No need to log here as the persistent task framework will log it
                                         try {
@@ -640,7 +643,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
         processContext.tryLock();
         try {
             if (processContext.setDying() == false) {
-                logger.debug("Cannot close job [{}] as it has already been closed", jobId);
+                logger.debug("Cannot close job [{}] as it has been marked as dying", jobId);
                 return;
             }
 

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
@@ -434,6 +434,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
                             return;
                         }
 
+                        processByAllocation.putIfAbsent(jobTask.getAllocationId(), new ProcessContext(jobTask));
                         jobResultsProvider.getAutodetectParams(job, params -> {
                             // We need to fork, otherwise we restore model state from a network thread (several GET api calls):
                             threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(new AbstractRunnable() {
@@ -444,8 +445,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
 
                                 @Override
                                 protected void doRun() {
-                                    ProcessContext processContext = new ProcessContext(jobTask);
-
+                                    ProcessContext processContext = processByAllocation.get(jobTask.getAllocationId());
                                     if (processContext == null) {
                                         logger.debug("Aborted opening job [{}] as it has been closed", jobId);
                                         return;
@@ -460,9 +460,6 @@ public class AutodetectProcessManager implements ClusterStateListener {
                                         createProcessAndSetRunning(processContext, job, params, closeHandler);
                                         processContext.getAutodetectCommunicator().restoreState(params.modelSnapshot());
                                         setJobState(jobTask, JobState.OPENED);
-
-                                        assert processByAllocation.get(jobTask.getAllocationId()) == null;
-                                        processByAllocation.put(jobTask.getAllocationId(), processContext);
                                     } catch (Exception e1) {
                                         // No need to log here as the persistent task framework will log it
                                         try {

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java
@@ -458,7 +458,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
 
                                     try {
                                         createProcessAndSetRunning(processContext, job, params, closeHandler);
-                                        processContext.getAutodetectCommunicator().init(params.modelSnapshot());
+                                        processContext.getAutodetectCommunicator().restoreState(params.modelSnapshot());
                                         setJobState(jobTask, JobState.OPENED);
                                     } catch (Exception e1) {
                                         // No need to log here as the persistent task framework will log it
@@ -499,7 +499,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
     private void createProcessAndSetRunning(ProcessContext processContext,
                                             Job job,
                                             AutodetectParams params,
-                                            BiConsumer<Exception, Boolean> handler) {
+                                            BiConsumer<Exception, Boolean> handler) throws IOException {
         // At this point we lock the process context until the process has been started.
         // The reason behind this is to ensure closing the job does not happen before
         // the process is started as that can result to the job getting seemingly closed
@@ -507,6 +507,7 @@ public class AutodetectProcessManager implements ClusterStateListener {
         processContext.tryLock();
         try {
             AutodetectCommunicator communicator = create(processContext.getJobTask(), job, params, handler);
+            communicator.writeHeader();
             processContext.setRunning(communicator);
         } finally {
             // Now that the process is running and we have updated its state we can unlock.

